packages/fxa-settings/src/lib/integrations/integration-factory-flags.ts

```diff
@@ -0,0 +1,131 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { Constants } from '../constants';
+import { ModelDataStore, UrlQueryData } from '../model-data';
+import { IntegrationFlags } from './interfaces/integration-flags';
+
+const DEVICE_PAIRING_SUPPLICANT_PATHNAME_REGEXP = /^\/pair\/supp/;
+
+/**
+ * Extrapolates flags from the state of the current data store. The collective state of these flags are used by
+ * the factory to determine what underlying type of integration to create.
+ *
+ * Note: this logic was ported from fxa-content-server app-start.js.
+ */
+export class DefaultIntegrationFactoryFlags implements IntegrationFlags {
+  protected get pathname() {
+    return this.urlQueryData.pathName;
+  }
+  constructor(private urlQueryData: UrlQueryData) {}
+
+  // TODO: Same method on the Relier flags, can we DRY?
+  isDevicePairingAsAuthority() {
+    return (
+      this._searchParam('redirect_uri') ===
+      Constants.DEVICE_PAIRING_AUTHORITY_REDIRECT_URI
+    );
+  }
+
+  isDevicePairingAsSupplicant() {
+    return DEVICE_PAIRING_SUPPLICANT_PATHNAME_REGEXP.test(this.pathname);
+  }
+
+  // TODO: Same method on the Relier flags, can we DRY?
+  isOAuth() {
+    return (
+      !!(
+        this._searchParam('client_id') ||
+        // verification
+        this._isOAuthVerificationSameBrowser()
+      ) ||
+      !!this._isOAuthVerificationDifferentBrowser() ||
+      // any URL with 'oauth' in the path.
+      /oauth/.test(this.pathname)
+    );
+  }
+
+  isOAuthWebChannel() {
+    // TODO
+    return false;
+  }
+
+  isChromeAndroid() {
+    // TODO
+    return false;
+  }
+
+  isSyncService() {
+    return this._isService(Constants.SYNC_SERVICE);
+  }
+
+  isV3DesktopContext() {
+    return this._searchParam('context') === Constants.FX_DESKTOP_V3_CONTEXT;
+  }
+
+  isOAuthSuccessFlow() {
+    const status = /oauth\/success/.test(this.pathname);
+
+    let clientId = '';
+    if (status) {
+      const pathname = this.pathname.split('/');
+      clientId = pathname[pathname.length - 1];
```

Comment:
Not sure if this is relevant to anything really, but would this work/break if we had rogue trailing backslashes in the URL?
ie: something like "/oauth/success" vs "/oauth/success/"

```js
const pathname1 = new URL("https://github.com/mozilla/fxa/pull/15096/files").pathname;
const pathname2 = new URL("https://github.com/mozilla/fxa/pull/15096/files/").pathname;

console.log(pathname1.split("/").at(-1)); // "files"
console.log(pathname2.split("/").at(-1)); // ""
```